File: comparative_tests/go_tests/image_manipulation_functions/contrast_filter.go
package image_manipulation_functions

import (
	"fmt"
	"image"
	"image/color"
	"math"
	"time"
)

func ContrastFilter(img image.Image, contrastFactor float64, outputsPath string) {
	start := time.Now()

	// Ensure the output directory exists
	if err := EnsureOutputDir(outputsPath); err != nil {
		fmt.Println(err)
		return
	}

	bounds := img.Bounds()
	contrastAdjustedImage := image.NewRGBA(bounds)

	// Calculate the contrast adjustment
	contrast := (259 * (contrastFactor + 255)) / (255 * (259 - contrastFactor))

	// Apply the contrast filter
	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		for x := bounds.Min.X; x < bounds.Max.X; x++ {
			// Get the original pixel color
			originalColor := color.RGBAModel.Convert(img.At(x, y)).(color.RGBA)

			// Adjust the contrast
			r := truncate(contrast*(float64(originalColor.R)-128) + 128)
			g := truncate(contrast*(float64(originalColor.G)-128) + 128)
			b := truncate(contrast*(float64(originalColor.B)-128) + 128)

			// Set the new pixel color
			contrastAdjustedImage.Set(x, y, color.RGBA{R: r, G: g, B: b, A: originalColor.A})
		}
	}

	// Create the output file
	outputFile, err := CreateOutputFile(outputsPath)
	if err != nil {
		fmt.Println(err)
		return
	}

	// Encode the resized image as JPEG and save it to the output file
	if err := SaveImageAsJPEG(outputFile, contrastAdjustedImage); err != nil {
		fmt.Println(err)
		return
	}

	fmt.Println("Contrast adjusted image saved to:", outputsPath)
	fmt.Println("Execution time:", time.Since(start))
}

func truncate(value float64) uint8 {
	return uint8(math.Min(math.Max(value, 0), 255))
}

File: comparative_tests/python_tests/image_manipulation_functions/contrast_filter.py
from PIL import Image
import numpy as np
from . import utils
import time


def truncate(value):
    return max(0, min(int(value), 255))

def contrast_filter(img, contrast_factor, outputs_path):
    start_time = time.time()

    # Ensure the output directory exists
    utils.ensure_output_dir(outputs_path)

    # Convert image to numpy array
    img_array = np.array(img.convert("RGBA"))

    # Calculating the contrast factor
    contrast = (259 * (contrast_factor + 255)) / (255 * (259 - contrast_factor))

    # Apply the contrast adjustment
    contrast_adjusted_array = img_array.copy()
    for y in range(img_array.shape[0]):
        for x in range(img_array.shape[1]):
            # Adjusting the contrast for each channel
            r = truncate(contrast * (img_array[y, x, 0] - 128) + 128)
            g = truncate(contrast * (img_array[y, x, 1] - 128) + 128)
            b = truncate(contrast * (img_array[y, x, 2] - 128) + 128)

            # Replacing the pixel values
            contrast_adjusted_array[y, x] = [r, g, b, img_array[y, x, 3]]

    # Conversion of the numpy array to an image
    contrast_adjusted_img = Image.fromarray(contrast_adjusted_array, "RGBA")

    # Save the contrast adjusted image
    output_file = utils.create_output_file(outputs_path)
    utils.save_image_as_jpeg(output_file, contrast_adjusted_img.convert("RGB"))

    print("Contrast adjusted image saved to:", output_file)
    print("Execution time:", time.time() - start_time, "seconds")
